pkg/clients/cloudwatch/concurrency.go

```diff
@@ -0,0 +1,69 @@
+package cloudwatch
+
+// ConcurrencyConfig configures how concurrency should be limited in a Cloudwatch API client. It allows
+// one to pick between different limiter implementations: a single limit limiter, or one with a different limit per
+// API call.
+type ConcurrencyConfig struct {
+	// PerAPIEnabled configures wether to have a limit per API call.
+	PerAPIEnabled bool
+
+	// SingleLimit configures the concurrency limit when using a single limiter for api calls.
+	SingleLimit int
+
+	// ListMetrics limits the number for ListMetrics API concurrent API calls.
+	ListMetrics int
+
+	// GetMetricData limits the number for GetMetricData API concurrent API calls.
+	GetMetricData int
+
+	// GetMetricStatistics limits the number for GetMetricStatistics API concurrent API calls.
+	GetMetricStatistics int
+}
+
+// NewLimiter creates a new ConcurrencyLimiter, according to the ConcurrencyConfig.
+func (cfg ConcurrencyConfig) NewLimiter() ConcurrencyLimiter {
+	if cfg.PerAPIEnabled {
+		return NewPerAPICallLimiter(cfg.ListMetrics, cfg.GetMetricData, cfg.GetMetricStatistics)
+	}
+	return NewSingleLimiter(cfg.SingleLimit)
+}
+
+// perAPICallLimiter is a ConcurrencyLimiter that keeps a different concurrency limiter per different API call. This allows
+// a more granular control of concurrency, allowing us to take advantage of different api limits. For example, ListMetrics
+// has a limit of 25 TPS, while GetMetricData has none.
+type perAPICallLimiter map[string]chan struct{}
```

Comment:
Could we potentially use three different `singleLimiter` instead of the `perAPICallLimiter`? I think that would be a bit more readable perhaps?

Reply:
yes, completely

Reply:
ok so extracted the common funcionality into a simple semaphore type, and re-used that everywhere.